ethcore/private-tx/src/log.rs

```diff
@@ -0,0 +1,409 @@
+use serde::ser::{Serializer, SerializeSeq};
+use error::Error;
+
+#[cfg(not(time_checked_add))]
+use time_utils::CheckedSystemTime;
+
+/// Maximum amount of stored private transaction logs.
+const MAX_JOURNAL_LEN: usize = 1000;
+
+/// Maximum period for storing private transaction logs.
+/// Logs older than 20 days will not be processed
+const MAX_STORING_TIME: Duration = Duration::from_secs(60 * 60 * 24 * 20);
+
+/// Source of monotonic time for log timestamps
+struct MonoTime {
+	start_time: SystemTime,
+	start_inst: Instant
+}
+
+impl MonoTime {
+	fn new(start: SystemTime) -> Self {
+		Self {
+			start_time: start,
+			start_inst: Instant::now()
+		}
+	}
+
+	fn elapsed(&self) -> Duration {
+		self.start_inst.elapsed()
+	}
+
+	fn to_system_time(&self) -> SystemTime {
+		self.start_time + self.elapsed()
+	}
+}
+
+impl Default for MonoTime {
+	fn default() -> Self {
+		MonoTime::new(SystemTime::now())
+	}
+}
+
+/// Current status of the private transaction
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq)]
+pub enum PrivateTxStatus {
+	/// Private tx was created but no validation received yet
+	Created,
+	/// Several validators (but not all) validated the transaction
+	Validating,
+	/// All validators has validated the private tx
+	/// Corresponding public tx was created and added into the pool
+	Deployed,
+}
+
+/// Information about private tx validation
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ValidatorLog {
+	/// Account of the validator
+	pub account: Address,
+	/// Validation timestamp, None if the transaction is not validated
+	pub validation_timestamp: Option<SystemTime>,
+}
+
+#[cfg(test)]
+impl PartialEq for ValidatorLog {
+	fn eq(&self, other: &Self) -> bool {
+		self.account == other.account
+	}
+}
+
+/// Information about the private transaction
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct TransactionLog {
+	/// Original signed transaction hash (used as a source for private tx)
+	pub tx_hash: H256,
+	/// Current status of the private transaction
+	pub status: PrivateTxStatus,
+	/// Creation timestamp
+	pub creation_timestamp: SystemTime,
+	/// List of validations
+	pub validators: Vec<ValidatorLog>,
+	/// Timestamp of the resulting public tx deployment
+	pub deployment_timestamp: Option<SystemTime>,
+	/// Hash of the resulting public tx
+	pub public_tx_hash: Option<H256>,
+}
+
+#[cfg(test)]
+impl PartialEq for TransactionLog {
+	fn eq(&self, other: &Self) -> bool {
+		self.tx_hash == other.tx_hash &&
+		self.status == other.status &&
+		self.validators == other.validators &&
+		self.public_tx_hash == other.public_tx_hash
+	}
+}
+
+/// Wrapper other JSON serializer
+pub trait LogsSerializer: Send + Sync + 'static {
+	/// Read logs from the source
+	fn read_logs(&self) -> Result<Vec<TransactionLog>, Error>;
+
+	/// Write all logs to the source
+	fn flush_logs(&self, logs: &HashMap<H256, TransactionLog>) -> Result<(), Error>;
+}
+
+pub struct FileLogsSerializer {
```

Comment:
Missing docs, perhaps add `#![deny(missing_docs)]` in `src/lib.rs` to enforce it